design/design-doc-agent.md

```diff
@@ -0,0 +1,665 @@
+    `MySQLInstanceRequirement` in the remote cluster.
+* Operation experience. This is related to the first point. If you have an app
+  using OAM, or some other app model, then you always have that intermediate
+  proxy of workload CRs. You're losing on some value that these models provide
+  because of that proxy and in some cases it could be functionally detrimental.
+
+Surely, it has its own advantages as well. For example, you can manage all of
+your apps from single point via `KubernetesApplication`s targeting the right
+clusters. But as we see more usage patterns, we're convinced that the current
+mechanics do not provide the experience users would like to have.
+
+## Proposal
+
+In order to preserve the central infrastructure management ability while
+alleviating the issues above, we will change our approach from push-based one to
+a pull-based one where applications, and their requirements are deployed into
+the remote cluster, and they request the infrastructure from a central cluster
+and pull the necessary credentials.
+
+Since having this logic in the applications themselves wouldn't be a good UX, we
+will have an agent that you will need to deploy into your remote cluster for
+doing the heavy-lifting for you. There are several technical problems to be
+solved in order to make the experience smooth. Overall, the goal is that we want
+to keep the UX of local mode for application operators while keeping the power
+of centralized infrastructure management for platform operators. For reference,
+here is an example local mode experience we'd like to have for the remote mode
+as well:
+
+```yaml
+apiVersion: common.crossplane.io/v1alpha1
+kind: MySQLInstanceRequirement
+metadata:
+  name: sqldb
+  namespace: default
+spec:
+  version: "5.7"
+  storageGB: 20
+  writeConnectionSecretToRef:
+    name: sql-creds
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wp
+  namespace: default
+spec:
+containers:
+- name: wordpress
+    image: "wordpress:4.6.1-apache"
+    env:
+    - name: WORDPRESS_DB_PASSWORD
+    valueFrom:
+        secretKeyRef:
+          name: sql-creds
+          key: password
+```
+
+The agent will be a Kubernetes controller running in the remote cluster and
+watching all `*Requirement` types. Next sections will talk about the
+implementation and user experience we'd like to have.
+
+### Synchronization
+
+In local mode, users directly interact with what's published by the platform
+team, which is `*Requirement` types and consume the infrastructure by mounting
+the secret whose name they specified on the `*Requirement` custom resource. To
+keep this experience, we need to have a synchronization loop for the following
+resources:
+
+* Pull
+  * `CustomResourceDefinition`s of all types that we want the applications to be
+    able to manage and view:
+    * Requirements that are published via `InfrastructurePublication`s. The
+      source of truth will be the remote cluster.
```

Comment:
Could you add a little context somewhere in this doc that explains why we'd pull the require CRDs from the central cluster rather than mirroring the logic that generates a CRD from an InfraPub in the agent? I'm not entirely convinced that it wouldn't be more flexible for the agent to mirror the logic (e.g. if we wanted the CRD gen process to diverge slightly at some point in future).

Reply:
The main reason is that I don't want to have 2 different source of truth for the resulting CRD, i.e. the agent and the Crossplane. For example, let's say Crossplane is updated to a new version in the central cluster but agents connected to the namespaces are still on an old version. Suddenly, CRDs could be different and sync could face unexpected errors because the schemas are different. Also by design, InfraPub and InfraDef resources are read-only in the remote cluster so there isn't much gain to have the generation in remote cluster as well because in the end, the schemas that are of importance are the ones in the central cluster.

Reply:
@negz Added a paragraph explaining the reasoning.